src/simple_cache_client.rs

```diff
@@ -61,44 +56,61 @@ pub struct SimpleCacheClientBuilder {
     data_channel: Channel,
     auth_token: String,
     default_ttl_seconds: NonZeroU64,
+    user_agent_name: String,
 }
 
 impl SimpleCacheClientBuilder {
-    pub async fn new(
+    /// Returns a builder which can produce an instance of a Momento client
+    ///
+    /// # Arguments
+    ///
+    /// * `auth_token` - Momento Token
+    /// * `item_default_ttl_seconds` - Default TTL for items put into a cache.
+    /// # Examples
+    ///
+    /// ```
+    /// # tokio_test::block_on(async {
+    ///     use momento::simple_cache_client::SimpleCacheClientBuilder;
+    ///     use std::env;
+    ///     use std::num::NonZeroU64;
+    ///     let auth_token = env::var("TEST_AUTH_TOKEN").expect("TEST_AUTH_TOKEN must be set");
+    ///     let default_ttl = 30;
+    ///     let momento = SimpleCacheClientBuilder::new(auth_token, NonZeroU64::new(default_ttl).unwrap())
+    ///         .expect("could not create a client")
+    ///         .build();
+    /// # })
+    /// ```
+    pub fn new(auth_token: String, default_ttl_seconds: NonZeroU64) -> Result<Self, MomentoError> {
+        SimpleCacheClientBuilder::new_with_explicit_agent_name(
+            auth_token,
+            default_ttl_seconds,
+            "sdk",
+        )
+    }
+
+    /// Like new() above, but used for naming integrations.
+    pub fn new_with_explicit_agent_name(
         auth_token: String,
         default_ttl_seconds: NonZeroU64,
+        user_agent_name: &str,
     ) -> Result<Self, MomentoError> {
-        let data_endpoint = match decode_jwt(&auth_token) {
-            Ok(claims) => claims.c,
-            Err(e) => return Err(e),
-        };
-
         let momento_endpoints = match MomentoEndpointsResolver::resolve(&auth_token, &None) {
             Ok(endpoints) => endpoints,
             Err(e) => return Err(e),
         };
-        let control_endpoint_uri = Uri::try_from(&momento_endpoints.control_endpoint)?;
-        let data_endpoint_uri = Uri::try_from(&momento_endpoints.data_endpoint)?;
-
-        let control_channel = Channel::builder(control_endpoint_uri)
-            .tls_config(ClientTlsConfig::default())
-            .unwrap()
-            .connect()
-            .await?;
+        log::debug!("connecting to endpoints: {:?}", momento_endpoints);
 
-        let data_channel = Channel::builder(data_endpoint_uri)
-            .tls_config(ClientTlsConfig::default())
-            .unwrap()
-            .connect()
-            .await?;
+        let control_channel = connect_channel_lazily(&momento_endpoints.control_endpoint)?;
+        let data_channel = connect_channel_lazily(&momento_endpoints.data_endpoint)?;
 
         match utils::is_ttl_valid(&default_ttl_seconds) {
             Ok(_) => Ok(Self {
-                data_endpoint,
+                data_endpoint: momento_endpoints.data_endpoint,
```

Comment:
We include in the create_signing_key response the data endpoint as is from the auth token (i.e. not `momento_endpoints.data_endpoint`, which added `"https://{}:443"`). The signing key integ test should fail with this code change. 
To avoid parsing auth token twice we can update `MomentoEndpointsResolver::resolve` to take a parsed claim instead of the token?

Reply:
I've reworked MomentoEndpointsResolver and removed a bunch of copypasta. That should pass now, thanks for the note.